KCL book, second edition (#29)
KittyCAD/kcl-book: kcl-book/src/fillets.md

# Fillets, Chamfers and Edges

<!-- toc -->

## Motivation: Applying a fillet

When you manufacture a part, you often want to smooth off its sharp edges, so they're rounded and won't accidentally cut someone who holds it.
Let's say we're modeling a cube, like this:

```kcl=cube_no_fillets
// Sketch a square
width = 1
square = sketch(on = XY) {
  line1 = line(start = [width / 2, -width / 2], end = [width / 2, width / 2])
  line2 = line(start = [width / 2, width / 2], end = [-width / 2, width / 2])
  line3 = line(start = [-width / 2, width / 2], end = [-width / 2, -width / 2])
  line4 = line(start = [-width / 2, -width / 2], end = [width / 2, -width / 2])
}

// Extrude a cube.
regionCube = region(point = [0.4975mm, 0mm], sketch = square)
extrudeCube = extrude(regionCube, length = width)
```

It produces a cube like this:

<!-- KCL: name=cube_no_fillets,alt=A cube -->

> What if we want to fillet one of its sides? Let's start simple and refer to one of the four bottom edges. Those edges were made by the four `line` calls, which were all assigned to variables. When we extruded the square into a cube, the variables were copied into the solid, under `.sketch.tags`. So we can reference the edge created from `line1` via `.sketch.tags.line1`, and apply a fillet to it.

Reviewer: ```suggestion
What if we want to fillet one of its sides? Let's start simple and refer to one of the four bottom edges. Those edges were made by the four `line` function calls, which were all assigned to variables, `line1`, `line2` etc. When we extruded the square into a cube, the variables were copied into the solid, under `.sketch.tags`. So we can reference the edge created from `line1` via `.sketch.tags.line1`, and apply a fillet to it.
```

```kcl=cube_one_fillet
// Sketch a square
width = 1
square = sketch(on = XY) {
  line1 = line(start = [width / 2, -width / 2], end = [width / 2, width / 2])
  line2 = line(start = [width / 2, width / 2], end = [-width / 2, width / 2])
  line3 = line(start = [-width / 2, width / 2], end = [-width / 2, -width / 2])
  line4 = line(start = [-width / 2, -width / 2], end = [width / 2, -width / 2])
}

// Extrude a cube.
regionCube = region(point = [0.4975mm, 0mm], sketch = square)
extrudeCube = extrude(regionCube, length = width)

// Fillet one edge
filletCube = fillet(extrudeCube, tags = extrudeCube.sketch.tags.line1, radius = 0.2)
```

The [`fillet`] function accepts an argument `tags`, which expects edges to fillet. You can pass in a single edge, like we did, or an array of edges like `[extrudeCube.sketch.tags.line1, extrudeCube.sketch.tags.line2]`.

That program should produce a cube with one filleted edge, like this:

<!-- KCL: name=cube_one_fillet,alt=A cube with one filleted edge -->

> Nice! We could fillet all four sides if we wanted to:

Reviewer: ```suggestion
Nice! We could fillet all four bottom sides if we wanted to:
```

```kcl=cube_four_fillets
// Sketch a square
width = 1
square = sketch(on = XY) {
  line1 = line(start = [width / 2, -width / 2], end = [width / 2, width / 2])
  line2 = line(start = [width / 2, width / 2], end = [-width / 2, width / 2])
  line3 = line(start = [-width / 2, width / 2], end = [-width / 2, -width / 2])
  line4 = line(start = [-width / 2, -width / 2], end = [width / 2, -width / 2])
}

// Extrude a cube.
regionCube = region(point = [0.4975mm, 0mm], sketch = square)
extrudeCube = extrude(regionCube, length = width)

> // Fillet one edge

Reviewer: ```suggestion
// Fillet all bottom edges
```

filletCube = fillet(
  extrudeCube,
  tags = [
    extrudeCube.sketch.tags.line1,
    extrudeCube.sketch.tags.line2,
    extrudeCube.sketch.tags.line3,
    extrudeCube.sketch.tags.line4,
  ],
  radius = 0.2,
)
```

<!-- KCL: name=cube_four_fillets,alt=A cube with four filleted edges-->

## Relationships between edges

> So far, we've assigned geometry (like a line) to a variable when we create it, and then use that variable to refer to it later (for fillets). What about edges we don't create directly, and therefore can't assign to a variable? For example, we've already filleted the four bottom edges, but how do we fillet the top four edges? We aren't creating them via `line` calls. They're created by the CAD engine in the `extrude` call. If we didn't explicitly create them with a sketch function, how do we store them in a variable? Here's the secret --- you don't. KCL has a few helpful functions to access edges that you didn't create directly. Because we can refer to the bottom edges, we can use helper functions like [`getOppositeEdge`] to reference the top edges, like this:

Reviewer: ```suggestion
So far, we've assigned geometry (like a line) to a variable when we create it, and then use that variable to refer to it later (e.g., for fillets). What about edges we don't create directly, and therefore can't assign to a variable? For example, we've already filleted the four bottom edges, but how do we fillet the top four edges? We aren't creating them via `line` calls. They're created by the CAD engine in the `extrude` call. If we didn't explicitly create them with a sketch function, how do we store them in a variable? Here's the secret --- you don't. KCL has a few helpful functions to access edges that you didn't create directly. Because we can refer to the bottom edges, we can use helper functions like [`getOppositeEdge`] to reference the top edges, like this:
```

## Measuring geometry

So we've learned to use variables from sketch blocks to reference the lines we create, then use helper functions like [`getOppositeEdge`] to reference other geometry elsewhere in the model. But these variables aren't just used for altering edges. They provide a valuable way to query and measure your models. Let's see how.

Let's say you've got a solid triangle, like this:

```kcl
// Make a triangle
sketch001 = sketch(on = YZ) {
  line1 = line(start = [var 5.29mm, var -4.11mm], end = [var -4.31mm, var -4.11mm])
  line2 = line(start = [var -4.31mm, var -4.11mm], end = [var 0.49mm, var 5.14mm])
  coincident([line1.end, line2.start])
  line3 = line(start = [var 0.49mm, var 5.14mm], end = [var 5.29mm, var -4.11mm])
  coincident([line2.end, line3.start])
  coincident([line3.end, line1.start])
  equalLength([line2, line3])
  horizontal(line1)
}

// Extrude it
region001 = region(point = [0.49mm, -4.1075mm], sketch = sketch001)
extrude001 = extrude(region001, length = 1)
```

Let's ask a simple question. How long is each side of the triangle?

> It sounds simple, but to actually calculate it, you'd have to break out a pencil and paper, then do some trigonometry. The problem is, the length doesn't appear anywhere in the `line` call. The lines are defined by their start and end points, and the length is an implicit property of those. Defining lines as a start and end is helpful, but it means important properties, like length, can't be read from our source code.

Reviewer: ```suggestion
It sounds simple, but to actually calculate it, you'd have to break out a pencil and paper, then do some trigonometry. The problem is, the length doesn't appear anywhere in the `line` function call. The lines are defined by their start and end points, and the length is an implicit property of those. Defining lines as a start and end is helpful, but it means important properties, like length, can't be read from our source code.
```

However, tags give us a simple way to refer to each line, and then query them for properties like length with the [`segLen`] function. Let's update our program:

```kcl
// Make a triangle
sketch001 = sketch(on = YZ) {
  line1 = line(start = [var 5.29mm, var -4.11mm], end = [var -4.31mm, var -4.11mm])
  line2 = line(start = [var -4.31mm, var -4.11mm], end = [var 0.49mm, var 5.14mm])
  coincident([line1.end, line2.start])
  line3 = line(start = [var 0.49mm, var 5.14mm], end = [var 5.29mm, var -4.11mm])
  coincident([line2.end, line3.start])
  coincident([line3.end, line1.start])
  equalLength([line2, line3])
  horizontal(line1)
}
// Extrude it
region001 = region(point = [0.49mm, -4.1075mm], sketch = sketch001)
extrude001 = extrude(region001, length = 1)

// Measure its side lengths

> side1Len = segLen(extrude001.sketch.tags.line1)

Reviewer: Unrelated to the book. 

This reads odd to me in code. We seem to use `sketch.tags.<var>` to get a face but in reality we are using the `<var>` as an "anchor" of sorts. The call to `segLen` reads to me as 

- we get the face/region created from `line1` and then we get the length on the line? 

We seem to conflate terms. Why can't I do 

`segLen(line1)` 

Also the function is called `segLen` which I assume expands to `segmentLength` so now we unified all these terms (an enum) under segment. I should check what we have in KCL in terms of types for these terms.

Author: I think we should be able to do `segLen(sketch001.line1)` (you can't use `line1` on its own, it is a child variable of the sketch block `sketch001`. It just doesn't work for now.

Author: Filed https://github.com/KittyCAD/modeling-app/pull/11350 for this.

side2Len = segLen(extrude001.sketch.tags.line2)
side3Len = segLen(extrude001.sketch.tags.line3)
```

> Now you can open up the Variables pane and look at the `lenA`, `lenB` and `lenC` variables to find each side's length. That's pretty useful! And if you want to use those lengths elsewhere in your code, you can! You could start drawing lines where the end is `[lenA, 0]` for example, or plug those lengths into other calculations. 

Reviewer: ```suggestion
Now you can open up the Variables pane and look at the `side1Len`, `side2Len` and `side3Len` variables to find each side's length. That's pretty useful! And if you want to use those lengths elsewhere in your code, you can! You could start drawing lines where the end is `[side1Len, 0]` for example, or plug those lengths into other calculations. 
```

> KCL has several other helper functions, like [`segAng`], which helps you find the angle between two lines. Let's measure the angles in a right-angle triangle:

Reviewer: This is another topic of confusion for me and the KCL code seems to carry over that confusion. An angle typically is defined between 2 lines. In math even its is the 3 names of the endpoints that form the angle ABC with the middle one being the shared endpoint. 

The function grabs a line/segment/face/region and comes up with the angle. What if it is the corner of a cube-like element. Which angle am I getting?

Author: Yeah that's fair. It works only on segments that form a closed profile, and it measures the angle between this line and the previous one. But it's a weird function. I'm going to remove it from these examples, because we should make one that actually works between pairs of lines like you describe.
